src/aws_encryption_sdk/keyring/multi_keyring.py

```diff
@@ -0,0 +1,107 @@
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Resources required for Multi Keyrings."""
+import itertools
+
+import attr
+from attr.validators import deep_iterable, instance_of, optional
+
+from aws_encryption_sdk.exceptions import EncryptKeyError, GenerateKeyError
+from aws_encryption_sdk.keyring.base import DecryptionMaterials, EncryptedDataKey, EncryptionMaterials, Keyring
+
+try:  # Python 3.5.0 and 3.5.1 have incompatible typing modules
+    from typing import Iterable  # noqa pylint: disable=unused-import
+except ImportError:  # pragma: no cover
+    # We only actually need these imports when running the mypy checks
+    pass
+
+
+@attr.s
+class MultiKeyring(Keyring):
+    """Public class for Multi Keyring.
+
+    :param generator: Generator keyring used to generate data encryption key (optional)
+    :type generator: Keyring
+    :param list children: List of keyrings used to encrypt the data encryption key (optional)
+    :raises EncryptKeyError: if encryption of data key fails for any reason
+    """
+
+    children = attr.ib(
+        default=attr.Factory(tuple), validator=optional(deep_iterable(member_validator=instance_of(Keyring)))
+    )
+    generator = attr.ib(default=None, validator=optional(instance_of(Keyring)))
+
+    def __attrs_post_init__(self):
+        # type: () -> None
+        """Prepares initial values not handled by attrs."""
+        neither_generator_nor_children = self.generator is None and not self.children
+        if neither_generator_nor_children:
+            raise TypeError("At least one of generator or children must be provided")
+
+        _generator = (self.generator,) if self.generator is not None else ()
+        self._decryption_keyrings = itertools.chain(_generator, self.children)
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        """Generate a data key using generator keyring
+        and encrypt it using any available wrapping key in any child keyring.
+
+        :param encryption_materials: Encryption materials for keyring to modify.
+        :type encryption_materials: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        :returns: Optionally modified encryption materials.
+        :rtype: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        :raises EncryptKeyError: if unable to encrypt data key.
+        """
+        # Check if generator keyring is not provided and data key is not generated
+        if self.generator is None and not encryption_materials.data_encryption_key:
```

Comment:
I missed this before. The unset value for `CryptographicMaterials.data_encryption_key` is `None`[1]. We should make sure that we are checking for `None` correctly everywhere we check it.

```suggestion
        if self.generator is None and encryption_materials.data_encryption_key is None:
```

[1] https://github.com/aws/aws-encryption-sdk-python/blob/keyring/src/aws_encryption_sdk/materials_managers/__init__.py#L91-L93